SimExLite/PMICalculators/SimpleScatteringPMICalculator.py

```diff
@@ -0,0 +1,11147 @@
+    #    data_prop_out.sel_pix_y = ( data_prop_out.ny + 1 ) / 2 ;
+
+    ##############################################################################
+
+    def f_time_evolution(self):
+
+        for step in range(1, self.g_s2e["steps"] + 1):
+            self.f_save_snp(step)
+
+    ##############################################################################
+
+
+def f_eval_disp(a_snp, a_r0, a_sample):
+
+    num_Z = len(list(a_sample["selZ"].keys()))
+    all_disp = numpy.zeros((num_Z,))
+    cc = 0
+    for sel_Z in list(a_sample["selZ"].keys()):
+        dr = a_snp["r"][a_sample["selZ"][sel_Z], :] - a_r0[a_sample["selZ"][sel_Z], :]
+        all_disp[cc] = numpy.mean(numpy.sqrt(numpy.sum(dr * dr, axis=1))) / 1e-10
+        cc = cc + 1
+    return all_disp
+
+
+def s2e_gen_randrot_quat(quat, rotmat):
+
+    if (
+        0
+        == quat[0] * quat[0] + quat[1] * quat[1] + quat[2] * quat[2] + quat[3] * quat[3]
+    ):
+        u0, u1, u2 = numpy.random.random(3)
+        q0 = numpy.sqrt(1.0 - u0) * numpy.sin(2 * numpy.pi * u1)
+        q1 = numpy.sqrt(1.0 - u0) * numpy.cos(2 * numpy.pi * u1)
+        q2 = numpy.sqrt(u0) * numpy.sin(2 * numpy.pi * u2)
+        q3 = numpy.sqrt(u0) * numpy.cos(2 * numpy.pi * u2)
+
+        quat[0] = q0
+        quat[1] = q1
+        quat[2] = q2
+        quat[3] = q3
+
+    else:
+        q0 = quat[0]
+        q1 = quat[1]
+        q2 = quat[2]
+        q3 = quat[3]
+
+    rotmat[0] = q0 * q0 + q1 * q1 - q2 * q2 - q3 * q3
+    # // M[0][0]
+    rotmat[1] = 2 * (q1 * q2 - q0 * q3)
+    # // M[0][1]
+    rotmat[2] = 2 * (q1 * q3 + q0 * q2)
+    # // M[0][2]
+    rotmat[3] = 2 * (q1 * q2 + q0 * q3)
+    # // M[1][0]
+    rotmat[4] = q0 * q0 - q1 * q1 + q2 * q2 - q3 * q3
+    # // M[1][1]
+    rotmat[5] = 2 * (q2 * q3 - q0 * q1)
+    # // M[1][2]
+    rotmat[6] = 2 * (q1 * q3 - q0 * q2)
+    # // M[2][0]
+    rotmat[7] = 2 * (q2 * q3 + q0 * q1)
+    # // M[2][1]
+    rotmat[8] = q0 * q0 - q1 * q1 - q2 * q2 + q3 * q3
+    # // M[2][2]
+
+
+#        inversematrix3( (void*) (rotmat + ii) , (void*) (invmat + ii) ) ;
+
+#    for ( ii = 0 ; ii < randrot ; ii++ )
+#    {
+#        for ( jj=0 ; jj < 3 ; jj++ )
+#        {
+#            printf( "% e % e % e\n" , rotmat[ii][jj*3+0] ,  rotmat[ii][jj*3+1] ,
+#                                      rotmat[ii][jj*3+2] ) ;
+#        }
+#        printf("\n") ;
+
+#        for ( jj=0 ; jj < 3 ; jj++ )
+#        {
+#            printf( "% e % e % e\n" , invmat[ii][jj*3+0] ,  invmat[ii][jj*3+1] ,
+#                                      invmat[ii][jj*3+2] ) ;
+#        }
+#        printf("\n\n\n") ;
+#    }
+# }
+
+
+##############################################################################
+
+
+def s2e_rand_orient(r, mat):
+
+    N = r.shape[0]
+    ###    print N
+    vv = numpy.zeros((3, 0))
+    for ii in range(N):
+        vv = r[ii, :]
+        ###        print vv , vv.shape , mat.shape
+        r[ii, 0] = mat[0] * vv[0] + mat[1] * vv[1] + mat[2] * vv[2]
```

Comment:
This could be a matrix multiplication, as simple as 
```
    for ii in range(N):
        vv = r[ii,:]
        r[ii] = mat.reshape(3,3) @ vv
```
Test this before usage though, I didn't use it much so may be wrong.